plutus-contract/src/Plutus/Contract/Test/ContractModel.hs

```diff
@@ -0,0 +1,1015 @@
+-- is required. Contract handles are managed behind the scenes by the `propRunActions` functions,
+-- based on a given a list of `ContractInstanceSpec`s, associating `ContractInstanceKey`s with `Wallet`s and
+-- `Contract`s. Before testing starts, `activateContractWallet` is called for all entries in the
+-- list and the mapping from `ContractInstanceKey` to `ContractHandle` is provided in the `HandleFun` argument
+-- to `perform`.
+
+-- | The constraints required on contract schemas and error types to enable calling contract
+--   endpoints (`Plutus.Trace.Emulator.callEndpoint`).
+type SchemaConstraints w schema err =
+        ( Typeable w
+        , Monoid w
+        , JSON.ToJSON w
+        , Typeable schema
+        , HasBlockchainActions schema
+        , ContractConstraints schema
+        , Show err
+        , Typeable err
+        , JSON.ToJSON err
+        , JSON.FromJSON err
+        )
+
+-- | A `ContractInstanceSpec` associates a `ContractInstanceKey` with a concrete `Wallet` and
+--   `Contract`. The contract type parameters are hidden from the outside.
+data ContractInstanceSpec state where
+    ContractInstanceSpec :: SchemaConstraints w schema err
+                  => ContractInstanceKey state w schema err -- ^ The key used when looking up contract instance handles in `perform`
+                  -> Wallet                                 -- ^ The wallet who owns the contract instance
+                  -> Contract w schema err ()               -- ^ The contract that is running in the instance
+                  -> ContractInstanceSpec state
+
+type Handles state = IMap (ContractInstanceKey state) ContractHandle
+
+
+-- | A function returning the `ContractHandle` corresponding to a `ContractInstanceKey`. A
+--   `HandleFun` is provided to the `perform` function to enable calling contract endpoints with
+--   `Plutus.Trace.Emulator.callEndpoint`.
+type HandleFun state = forall w schema err. (Typeable w, Typeable schema, Typeable err) => ContractInstanceKey state w schema err -> ContractHandle w schema err
+
+-- | The `ModelState` models the state of the blockchain. It contains,
+--
+--   * the contract-specific state (`contractState`)
+--   * the current slot (`currentSlot`)
+--   * the wallet balances (`balances`)
+--   * the amount that has been forged (`forged`)
+data ModelState state = ModelState
+        { _currentSlot   :: Slot
+        , _lastSlot      :: Slot
+        , _balances      :: Map Wallet Value
+        , _forged        :: Value
+        , _contractState :: state
+        }
+
+dummyModelState :: state -> ModelState state
+dummyModelState s = ModelState 0 0 Map.empty mempty s
+
+instance Show state => Show (ModelState state) where
+    show = show . _contractState   -- for now
+
+-- | The `Spec` monad is a state monad over the `ModelState`. It is used exclusively by the
+--   `nextState` function to model the effects of an action on the blockchain.
+newtype Spec state a = Spec (Eff '[State (ModelState state)] a)
```

Comment:
Agree. I went with `Eff` at first, since everything was using it, but since then more monad transformers have made it into the code. That said, I think there's a point in not letting the user manipulate the model state too freely.

Reply:
Is there? What kind of mistake do we think people would make if they could modify it directly? They can already modify almost every part of it...

Reply:
I'm changing it to state transformer and exposing a `MonadState` instance. I still think it's worth keeping the custom operators though, since they save you from a `contractState` projection, and if you're not fluent in `Lens` it might not be obvious how to use

```haskell
(%%=) :: MonadState s m => Over p ((,) r) s s a b -> p a (r, b) -> m r
```

to get the update you want.